README.md
# Improved Research Queue

This is a mod for the game [Factorio](https://www.factorio.com/). It can be downloaded from the [Factorio Mod Portal](https://mods.factorio.com/mod/sonaxaton-research-queue).

## Description

This mod adds a new GUI for managing a queue of research that's easier to use and more featured than the vanilla research queue.

* Maintain a research queue of unlimited size
* Move technologies up and down in the queue
* Adding an unavailable technology to the queue will add all of its prerequisites as well
* Removing a technology from the queue will remove everything that has it as a prerequisite as well
* Easily search and filter technologies to research
* Queue up individual levels of infinite technologies
* Pause the research queue while you plan, then unpause and watch your queue go
* See estimated time to completion and total number of science packs needed for technologies in the queue
* Receive a chat notification when a research finishes (can turn off in mod settings)
* Works with any modded technologies and science packs
* In multiplayer, players on the same team will share the same research queue
* Option to pause the game while you plan your research (off by default, turn on in mod settings)

This mod is intended to replace these other great but outdated mods:

* [Research queue](https://mods.factorio.com/mod/research-queue) / [Research Queue: The Old New Thing](https://mods.factorio.com/mod/research-queue-the-old-new-thing) / [Research Queue: The Old New Thing, 0.18.34 fix](https://mods.factorio.com/mod/research-queue-0-18-34)
* [Deadlock's Research Notifications](https://mods.factorio.com/mod/DeadlockResearchNotifications)

## How to Use

To open the Improved Research Queue window, use the shortcut on your shortcut bar or press `Shift+T`. You will see a list of technologies in the center, the queue on the left, and settings/filters on the right. Hover over any element to see additional info and descriptions of how to use it.

## Settings

Under `Settings` -> `Mod Settings` -> `Per player` -> `Improved Research Queue` you can customize the following settings:

* **Research completion notifications** - If enabled, a message will be posted in your chat when a research completes for your team. Helpful to be able to see a history of completed researches in the chat.
* **Pause game while window is open** - Pause the game while the Research Queue mod window is open. Only works in single player.

You can customize the the keyboard shortcut to open the Research Queue window in `Settings` -> `Controls` -> `Mods` -> `Improved Research Queue`.

## Tips

**Do not use the vanilla research queue with this mod**; this mod will most likely overwrite the vanilla queue. However, starting and cancelling research using the vanilla technology GUI will work fine with this mod. Starting a new research will add it to the start of the queue and unpause the queue. Cancelling the current research will pause the queue.

While using this mod you probably want to turn off the vanilla setting that brings up the technology screen after every research since the mod will auto-select the next research for you if you have a queue set up. Disable the option under `Settings` -> `Interface` -> `Interaction` -> `Technology screen opens when research is completed`.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.4.22
Date: 2022-03-08
  Changes:
    - interface for other mods added
---------------------------------------------------------------------------------------------------
Version: 0.4.21
Date: 2021-02-16
  Minor Changes:
    - Hide description in tooltip for technologies without a description translation.
---------------------------------------------------------------------------------------------------
Version: 0.4.20
Date: 2021-02-15
  Bugfixes:
    - Fix crash in some scenarios when research events are triggered before a player is initialized.
---------------------------------------------------------------------------------------------------
Version: 0.4.19
Date: 2021-02-14
  Bugfixes:
    - Fix crash when a technology has an infinite technology with no max level as a prerequisite. A technology like this would be impossible to research anyway, but at least now Improved Research Queue will just log a warning instead of crashing.
    - Fix layout issue in technology list at UI scales other than 100%.
---------------------------------------------------------------------------------------------------
Version: 0.4.18
Date: 2021-01-15
  Bugfixes:
    - Fix desync related to translation handler.
---------------------------------------------------------------------------------------------------
Version: 0.4.17
Date: 2021-01-15
  Bugfixes:
    - Fix desync related to translation handler.
  Minor Changes:
    - Clean up legacy flib data
---------------------------------------------------------------------------------------------------
Version: 0.4.16
Date: 2021-01-05
  Minor Changes:
    - Significant performance improvements when using large mods
---------------------------------------------------------------------------------------------------
Version: 0.4.15
Date: 2020-12-26
  Bugfixes:
    - Fix crash on Factorio 1.0 when auto-selecting science packs due to using API added in Factorio 1.1
---------------------------------------------------------------------------------------------------
Version: 0.4.14
Date: 2020-12-24
  Minor Changes:
    - Releasing Factorio 1.0-compatible version
---------------------------------------------------------------------------------------------------
Version: 0.4.13
Date: 2020-12-17
  Bugfixes:
    - Fix crash when loading certain modded infinite technologies
---------------------------------------------------------------------------------------------------
Version: 0.4.12
Date: 2020-12-17
  Bugfixes:
    - Fix crash when researching certain modded infinite technologies
---------------------------------------------------------------------------------------------------
Version: 0.4.11
Date: 2020-12-16
  Features:
    - Better support for multi-level and infinite technologies. If a technology has multiple levels or is infinite, you can now queue up individual levels as separate items in the queue.
  Minor Changes:
    - Performance improvements
---------------------------------------------------------------------------------------------------
Version: 0.4.10
Date: 2020-12-14
  Features:
    - Added technology description to tooltip
    - Added button to clear the queue
  Changes:
    - Infinite technologies will now be removed from the queue once researched instead of staying as the current research until you removed it manually. It's currently not possible to queue up more than one level of an infinite research at a time. I'm working on making this possible but it's quite a big change to make to the code so it might take a while.
  Bugfixes:
    - Fixed that science packs created by rocket launches wouldn't be auto-selected once researched
---------------------------------------------------------------------------------------------------
Version: 0.4.9
Date: 2020-12-12
  Features:
    - Added total number of science packs needed to research a technology in the queue to the ETC tooltip
  Bugfixes:
    - Fixed crash on load while saving queue for forces created by other mods
---------------------------------------------------------------------------------------------------
Version: 0.4.8
Date: 2020-12-04
  Bugfixes:
    - Fixed crash on new map when another completes a research while creating a force
---------------------------------------------------------------------------------------------------
Version: 0.4.7
Date: 2020-11-30
  Bugfixes:
    - Fixed crash when opening Krastorio2 wiki
---------------------------------------------------------------------------------------------------
Version: 0.4.6
Date: 2020-11-29
  Minor Changes:
    - Updated to Factorio 1.1
---------------------------------------------------------------------------------------------------
Version: 0.4.5
Date: 2020-11-09
  Features:
    - Added mod setting to pause the game while the window is open like the vanilla GUI does (off by default)
  Changes:
    - Made the window bigger
  Minor Changes:
    - Translate all terms used for search on mod startup instead of on-demand
---------------------------------------------------------------------------------------------------
Version: 0.4.4
Date: 2020-11-05
  Bugfixes:
    - Fix associated keybinding for shortcut button
---------------------------------------------------------------------------------------------------
Version: 0.4.3
Date: 2020-11-05
  Bugfixes:
    - Fix crash when a player is removed
---------------------------------------------------------------------------------------------------
Version: 0.4.2
Date: 2020-11-04
  Bugfixes:
    - Fix that some technologies were incorrectly considered "upgrade" technologies and hidden by default
---------------------------------------------------------------------------------------------------
Version: 0.4.1
Date: 2020-11-04
  Bugfixes:
    - Disable debug features accidentally left on
---------------------------------------------------------------------------------------------------

prototypes/styles.lua
local util = require('util')

local styles = data.raw['gui-style'].default

local tech_list_tech_button_size = 64+8*2+8
local tech_queue_tech_button_size = 64*4/4+8

local function tech_graphical_set(type, opts)
  local highlighted = opts.highlighted
  local y = ({
    available = 136,
    conditionally_available = 153,
    unavailable = 170,
    researched = 187,
    disabled = 204,
  })[opts.style]
  local tint = opts.tint

  if y == nil then
    error('unknown tech_graphical_set style')
  end

  if type == 'button' then
    local default_graphical_set = {
      base = {
        position = {highlighted and 330 or 296, y},
        corner_size = 8,
        tint = tint,
      },
      shadow = default_shadow,
    }
    local other_graphical_set = {
      base = {
        position = {312, y},
        corner_size = 8,
        tint = tint,
      },
      shadow = default_shadow,
    }
    return {
      default_graphical_set = default_graphical_set,
      hovered_graphical_set = other_graphical_set,
      clicked_graphical_set = other_graphical_set,
      selected_graphical_set = other_graphical_set,
      selected_hovered_graphical_set = other_graphical_set,
      selected_clicked_graphical_set = other_graphical_set,
    }
  elseif type == 'ingredients' then
    return {
      base = {
        position = {highlighted and 364 or 347, y},
        corner_size = 8,
        tint = tint,
      },
      shadow = default_shadow,
    }
  elseif type == 'level' then
    return {
      base = {
        position = {397, y},
        corner_size = 8,
        tint = tint,
      },
      shadow = default_shadow,
    }
  else
    error('unknown tech_graphical_set type '..serpent.line(type))
  end
end

styles.rq_main_window = {
  type = 'frame_style',
  parent = 'inner_frame_in_outer_frame',
  height = 500,
}

styles.rq_settings_window = {
  type = 'frame_style',
  parent = 'inner_frame_in_outer_frame',
  height = 500,
}

styles.rq_list_box = {
  type = 'scroll_pane_style',
  vertically_stretchable = 'on',
  extra_padding_when_activated = 0,
  graphical_set = {
    base = {position = {34, 0}, corner_size = 8},
    shadow = default_inner_shadow,
  },
}

styles.rq_tech_list_list_box = {
  type = 'scroll_pane_style',
  parent = 'rq_list_box',
  width = 4+5*tech_list_tech_button_size+4+12,
  background_graphical_set = {
    position = {282, 17},
    corner_size = 8,
    overall_tiling_horizontal_size = tech_list_tech_button_size-8,
    overall_tiling_horizontal_padding = 8/2,
    overall_tiling_horizontal_spacing = 8,
    overall_tiling_vertical_size = tech_list_tech_button_size+4+16+4+40-8,
    overall_tiling_vertical_padding = 8/2,
    overall_tiling_vertical_spacing = 8,
  },
}

styles.rq_tech_ingredient_filter_table_scroll_box = {
  type = 'scroll_pane_style',
}

styles.rq_tech_list_search_container = {
  type = 'horizontal_flow_style',
  horizontal_align = 'right',
  horizontally_stretchable = 'on',
  vertical_align = 'center',
}

styles.rq_filter_researched_button_enabled = {
  type = 'button_style',
  parent = 'green_slot',
  size = 28,
  padding = 2,
}

styles.rq_filter_researched_button_disabled = {
  type = 'button_style',
  parent = 'slot',
  size = 28,
  padding = 2,
}

styles.rq_tech_ingredient_filter_button_enabled = {
  type = 'button_style',
  parent = 'green_slot',
}

styles.rq_tech_ingredient_filter_button_disabled = {
  type = 'button_style',
  parent = 'slot',
}

styles.rq_tech_list_table = {
  type = 'table_style',
  horizontally_stretchable = 'off',
  horizontal_spacing = 0,
  vertical_spacing = 0,
}

styles.rq_tech_list_item = {
  type = 'vertical_flow_style',
  vertical_spacing = 0,
}

local tech_list_item_available = {style='available'}
local tech_list_item_unavailable = {style='unavailable'}
local tech_list_item_queued = {style='conditionally_available'}
local tech_list_item_queued_head = {style='conditionally_available', highlighted=true}
local tech_list_item_researched = {style='researched'}

styles.rq_tech_list_item_available_tech_button = util.merge{
  {
    type = 'button_style',
    size = tech_list_tech_button_size,
  },
  tech_graphical_set('button', tech_list_item_available),
}

styles.rq_tech_list_item_unavailable_tech_button = util.merge{
  {
    type = 'button_style',
    size = tech_list_tech_button_size,
  },
  tech_graphical_set('button', tech_list_item_unavailable),
}

styles.rq_tech_list_item_queued_tech_button = util.merge{
  {
    type = 'button_style',
    size = tech_list_tech_button_size,
  },
  tech_graphical_set('button', tech_list_item_queued),
}

styles.rq_tech_list_item_queued_head_tech_button = util.merge{
  {
    type = 'button_style',
    size = tech_list_tech_button_size,
  },
  tech_graphical_set('button', tech_list_item_queued_head),
}

styles.rq_tech_list_item_researched_tech_button = util.merge{
  {
    type = 'button_style',
    size = tech_list_tech_button_size,
  },
  tech_graphical_set('button', tech_list_item_researched),
}

styles.rq_tech_list_item_available_ingredients_bar = {
  type = 'frame_style',
  padding = 0,
  horizontal_flow_style = {
    type = 'horizontal_flow_style',
    horizontally_stretchable = 'on',
    horizontal_align = 'left',
  },
  graphical_set = tech_graphical_set('level', tech_list_item_available),
}

styles.rq_tech_list_item_unavailable_ingredients_bar = {
  type = 'frame_style',
  parent = 'rq_tech_list_item_available_ingredients_bar',
  graphical_set = tech_graphical_set('level', tech_list_item_unavailable),
}

styles.rq_tech_list_item_queued_ingredients_bar = {
  type = 'frame_style',
  parent = 'rq_tech_list_item_available_ingredients_bar',
  graphical_set = tech_graphical_set('level', tech_list_item_queued),
}

styles.rq_tech_list_item_queued_head_ingredients_bar = {
  type = 'frame_style',
  parent = 'rq_tech_list_item_available_ingredients_bar',
  graphical_set = tech_graphical_set('level', tech_list_item_queued_head),
}

styles.rq_tech_list_item_researched_ingredients_bar = {
  type = 'frame_style',
  parent = 'rq_tech_list_item_available_ingredients_bar',
  graphical_set = tech_graphical_set('level', tech_list_item_researched),
}

styles.rq_tech_list_item_ingredient = {
  type = 'image_style',
  width = 16,
  height = 16,
  stretch_image_to_widget_size = true,
}

styles.rq_tech_list_item_available_tool_bar = {
  type = 'frame_style',
  padding = 0,
  top_padding = 4,
  bottom_padding = 4,
  horizontal_flow_style = {
    type = 'horizontal_flow_style',
    horizontally_stretchable = 'on',
    horizontal_align = 'center',
  },
  graphical_set = tech_graphical_set('ingredients', tech_list_item_available),
}

styles.rq_tech_list_item_queued_tool_bar = {
  type = 'frame_style',
  parent = 'rq_tech_list_item_available_tool_bar',
  graphical_set = tech_graphical_set('ingredients', tech_list_item_queued),
}

styles.rq_tech_list_item_queued_head_tool_bar = {
  type = 'frame_style',
  parent = 'rq_tech_list_item_available_tool_bar',
  graphical_set = tech_graphical_set('ingredients', tech_list_item_queued_head),
}

styles.rq_tech_list_item_unavailable_tool_bar = {
  type = 'frame_style',
  parent = 'rq_tech_list_item_available_tool_bar',
  graphical_set = tech_graphical_set('ingredients', tech_list_item_unavailable),
}

styles.rq_tech_list_item_researched_tool_bar = {
  type = 'frame_style',
  parent = 'rq_tech_list_item_available_tool_bar',
  graphical_set = tech_graphical_set('ingredients', tech_list_item_researched),
}

styles.rq_tech_list_item_tool_button = {
  type = 'button_style',
  parent = 'tool_button',
  size = 24,
  padding = 0,
}

local tech_queue_item_inner_width = tech_queue_tech_button_size+8+16
local tech_queue_item_width = 2+4+tech_queue_item_inner_width+4+2
local tech_queue_item_inner_height = tech_queue_tech_button_size+4+20
local tech_queue_item_height = 2+4+tech_queue_item_inner_height+4+2

styles.rq_tech_queue_head_frame = {
  type = 'frame_style',
  width = 4+tech_queue_item_width+4,
  height = 4+tech_queue_item_height+4,
  padding = 0,
  horizontal_flow_style = {
    type = 'horizontal_flow_style',
    padding = 0,
  },
  graphical_set = {
    base = {position = {34, 0}, corner_size = 8},
    shadow = default_inner_shadow,
  },
  background_graphical_set = {
    position = {282, 17},
    corner_size = 8,
    overall_tiling_horizontal_size = tech_queue_item_width,
    overall_tiling_horizontal_padding = 4,
    overall_tiling_vertical_size = tech_queue_item_height,
    overall_tiling_vertical_padding = 4,
    overall_tiling_vertical_spacing = 4,
  },
}

styles.rq_tech_queue_item_paused = {
    type = 'horizontal_flow_style',
    horizontally_stretchable = 'on',
    -- FIXME: if this is vertically stretchable, the queue scrollbox gets shorter???
    height = tech_queue_item_height,
    -- vertically_stretchable = 'on',
    horizontal_align = 'center',
    vertical_align = 'center',
}

styles.rq_tech_queue_item_paused_unpause_button = {
  type = 'button_style',
  parent = 'tool_button_green',
}

styles.rq_tech_queue_list_box = {
  type = 'scroll_pane_style',
  parent = 'rq_list_box',
  width = 4+tech_queue_item_width+4+12,
  background_graphical_set = {
    position = {282, 17},
    corner_size = 8,
    overall_tiling_horizontal_size = tech_queue_item_width,
    overall_tiling_vertical_size = tech_queue_item_height,
    overall_tiling_vertical_spacing = 4,
  },
}

styles.rq_tech_queue_item = {
  type = 'frame_style',
  parent = 'subpanel_frame',
  padding = 2,
  horizontal_flow_style = {
    type = 'horizontal_flow_style',
    vertically_stretchable = 'off',
    horizontal_spacing = 8,
  },
}

styles.rq_tech_queue_item_inner_flow = {
  type = 'vertical_flow_style',
}

styles.rq_tech_queue_item_tech_button = util.merge{
  {
    type = 'button_style',
    size = tech_queue_tech_button_size,
    padding = 0,
  },
  tech_graphical_set('button', tech_list_item_queued),
}

styles.rq_tech_queue_head_item_tech_button = util.merge{
  {
    type = 'button_style',
    size = tech_queue_tech_button_size,
    padding = 0,
  },
  tech_graphical_set('button', tech_list_item_queued_head),
}

styles.rq_tech_queue_item_buttons = {
  type = 'vertical_flow_style',
  horizontal_align = 'center',
  vertical_spacing = 0,
  top_padding = 4,
  bottom_padding = 4,
}

styles.rq_tech_queue_item_close_button = {
  type = 'button_style',
  parent = 'mini_button',
}

styles.rq_tech_queue_item_shift_up_button = {
  type = 'button_style',
  size = {8, 8},
  padding = 0,
  default_graphical_set = {
    filename = '__core__/graphics/arrows/table-header-sort-arrow-up-active.png',
    size = {16, 16},
    scale = 0.5
  },
  hovered_graphical_set = {
    filename = '__core__/graphics/arrows/table-header-sort-arrow-up-hover.png',
    size = {16, 16},
    scale = 0.5
  },
  clicked_graphical_set = {
    filename = '__core__/graphics/arrows/table-header-sort-arrow-up-active.png',
    size = {16, 16},
    scale = 0.5
  },
  disabled_graphical_set = util.empty_sprite(),
}

styles.rq_tech_queue_item_shift_down_button = {
  type = 'button_style',
  size = {8, 8},
  padding = 0,
  default_graphical_set = {
    filename = '__core__/graphics/arrows/table-header-sort-arrow-down-active.png',
    size = {16, 16},
    scale = 0.5
  },
  hovered_graphical_set = {
    filename = '__core__/graphics/arrows/table-header-sort-arrow-down-hover.png',
    size = {16, 16},
    scale = 0.5
  },
  clicked_graphical_set = {
    filename = '__core__/graphics/arrows/table-header-sort-arrow-down-active.png',
    size = {16, 16},
    scale = 0.5
  },
  disabled_graphical_set = util.empty_sprite(),
}

styles.rq_etc_label = {
  type = 'label_style',
  parent = 'label',
  font = 'count-font',
  maximal_width = tech_queue_tech_button_size,
}

styles.rq_settings_section = {
  type = 'frame_style',
  parent = 'bordered_frame',
}

styles.rq_tech_button_container = {
  type = 'vertical_flow_style',
  vertical_spacing = 0,
  padding = 0,
  margin = 0,
}

styles.rq_tech_button_container_tech_list = {
  type = 'vertical_flow_style',
  parent = 'rq_tech_button_container',
  height = tech_list_tech_button_size,
}

styles.rq_tech_button_container_tech_queue = {
  type = 'vertical_flow_style',
  parent = 'rq_tech_button_container',
  height = tech_queue_tech_button_size,
}

styles.rq_tech_button_progressbar = {
  type = 'progressbar_style',
  horizontally_stretchable = 'on',
  bar_width = 4,
  bar_background = {position = {296, 39}, corner_size = 4},
  bar = {position = {305, 39}, corner_size = 4},
}

styles.rq_tech_button_progressbar_tech_list = {
  type = 'progressbar_style',
  parent = 'rq_tech_button_progressbar',
  top_margin = -tech_list_tech_button_size,
}

styles.rq_tech_button_progressbar_tech_queue = {
  type = 'progressbar_style',
  parent = 'rq_tech_button_progressbar',
  top_margin = -tech_queue_tech_button_size,
}

styles.rq_frame_action_button_green = {
  type = 'button_style',
  parent = 'frame_action_button',
  default_graphical_set = {
    base = {position = {420, 17}, corner_size = 8},
    shadow = {position = {440, 24}, corner_size = 8, draw_type = 'outer'},
  },
  hovered_graphical_set = {
    base = {position = {102, 17}, corner_size = 8},
    shadow = {position = {440, 24}, corner_size = 8, draw_type = 'outer'},
    glow = default_glow(green_button_glow_color, 0.5),
  },
  clicked_graphical_set = {
    base = {position = {119, 17}, corner_size = 8},
    shadow = {position = {440, 24}, corner_size = 8, draw_type = 'outer'},
  },
  disabled_graphical_set = {
    base = {position = {85, 17}, corner_size = 8},
    shadow = {position = {440, 24}, corner_size = 8, draw_type = 'outer'},
  },
}

styles.rq_frame_action_button_red = {
  type = 'button_style',
  parent = 'frame_action_button',
  default_graphical_set = {
    base = {position = {403, 17}, corner_size = 8},
    shadow = {position = {440, 24}, corner_size = 8, draw_type = 'outer'},
  },
  hovered_graphical_set = {
    base = {position = {170, 17}, corner_size = 8},
    shadow = {position = {440, 24}, corner_size = 8, draw_type = 'outer'},
    glow = default_glow(red_button_glow_color, 0.5),
  },
  clicked_graphical_set = {
    base = {position = {187, 17}, corner_size = 8},
    shadow = {position = {440, 24}, corner_size = 8, draw_type = 'outer'},
  },
  disabled_graphical_set = {
    base = {position = {153, 17}, corner_size = 8},
    shadow = {position = {440, 24}, corner_size = 8, draw_type = 'outer'},
  },
}

data:extend{
  {
    type = 'font',
    name = 'rq-notification',
    from = 'default-semibold',
    border = true,
    border_color = {},
    size = 18,
  },
}
